ksql-metastore/src/main/java/io/confluent/ksql/metastore/model/KeyField.java

```diff
@@ -224,4 +240,79 @@ public String toString() {
   private static boolean shouldUseLegacy(final KsqlConfig ksqlConfig) {
     return ksqlConfig.getBoolean(KsqlConfig.KSQL_USE_LEGACY_KEY_FIELD);
   }
+
+  @Immutable
+  public static final class LegacyField {
+
+    public final String name;
+    public final SqlType type;
+
+    /**
+     * This is used to replicate legacy logic to ensure unnecessary repartition steps are maintained
+     * going forward to ensure compatibility for existing queries.
+     */
+    private final boolean notInSchema;
+
+    public static LegacyField of(final String name, final SqlType type) {
+      return new LegacyField(name, type, false);
+    }
+
+    public static LegacyField notInSchema(final String name, final SqlType type) {
+      return new LegacyField(name, type, true);
+    }
+
+    private LegacyField(final String name, final SqlType type, final boolean notInSchema) {
+      this.name = Objects.requireNonNull(name, "name");
+      this.type = Objects.requireNonNull(type, "type");
+      this.notInSchema = notInSchema;
+    }
+
+    public String name() {
+      return name;
+    }
+
+    public SqlType type() {
+      return type;
+    }
+
+    public boolean isNotInSchema() {
```

Comment:
nit: it's probably cleaner to have `isInSchema` and invert the logic where need be (otherwise `!isNotInSchema` becomes a little unwieldily!)

while I'm here, what was our decision on backwards compatibility here? does this fit under the "let's just nuke this behavior because it's bad and causes undue code maintenance" category? I thought that was our general sentiment. though out of scope for this PR, we should consider just removing the LegacyField stuff

Reply:
Our general view is to maintain backwards compatibility where ever possible, and here it was possible.  Though I'm also game for removing this legacy code, and it was always my intention to do so, just not till the next major version bump.

Reply:
Regarding flipping `notInSchema`... I went to make this change only to find the only place `isNotInSchema` is used would have to become `if (!isInSchema())`. To my mind, if the only use would need to be negated if we flip, then we shouldn't flip it.